internal/maintenance/queue_cleaner.go

```diff
@@ -0,0 +1,162 @@
+	// QueueRetentionPeriod is the amount of time to keep queues around before
+	// they're removed.
+	QueueRetentionPeriod time.Duration
+}
+
+func (c *QueueCleanerConfig) mustValidate() *QueueCleanerConfig {
+	if c.Interval <= 0 {
+		panic("QueueCleanerConfig.Interval must be above zero")
+	}
+	if c.QueueRetentionPeriod <= 0 {
+		panic("QueueCleanerConfig.QueueRetentionPeriod must be above zero")
+	}
+
+	return c
+}
+
+// QueueCleaner periodically removes queues from the river_queue table that have
+// not been updated in a while, indicating that they are no longer active.
+type QueueCleaner struct {
+	queueMaintainerServiceBase
+	startstop.BaseStartStop
+
+	// exported for test purposes
+	Config      *QueueCleanerConfig
+	TestSignals QueueCleanerTestSignals
+
+	batchSize int // configurable for test purposes
+	exec      riverdriver.Executor
+}
+
+func NewQueueCleaner(archetype *baseservice.Archetype, config *QueueCleanerConfig, exec riverdriver.Executor) *QueueCleaner {
+	return baseservice.Init(archetype, &QueueCleaner{
+		Config: (&QueueCleanerConfig{
+			Interval:             valutil.ValOrDefault(config.Interval, QueueCleanerIntervalDefault),
+			QueueRetentionPeriod: valutil.ValOrDefault(config.QueueRetentionPeriod, QueueRetentionPeriodDefault),
+		}).mustValidate(),
+
+		batchSize: BatchSizeDefault,
+		exec:      exec,
+	})
+}
+
+func (s *QueueCleaner) Start(ctx context.Context) error {
+	ctx, shouldStart, stopped := s.StartInit(ctx)
+	if !shouldStart {
+		return nil
+	}
+
+	s.StaggerStart(ctx)
+
+	go func() {
+		// This defer should come first so that it's last out, thereby avoiding
+		// races.
+		defer close(stopped)
+
+		s.Logger.DebugContext(ctx, s.Name+logPrefixRunLoopStarted)
+		defer s.Logger.DebugContext(ctx, s.Name+logPrefixRunLoopStopped)
+
+		ticker := timeutil.NewTickerWithInitialTick(ctx, s.Config.Interval)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			res, err := s.runOnce(ctx)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					s.Logger.ErrorContext(ctx, s.Name+": Error cleaning queues", slog.String("error", err.Error()))
+				}
+				continue
+			}
+
+			s.Logger.InfoContext(ctx, s.Name+logPrefixRanSuccessfully,
+				slog.Int("num_queues_deleted", len(res.QueuesDeleted)),
+			)
+		}
+	}()
+
+	return nil
+}
+
+type queueCleanerRunOnceResult struct {
+	QueuesDeleted []string
+}
+
+func (s *QueueCleaner) runOnce(ctx context.Context) (*queueCleanerRunOnceResult, error) {
+	res := &queueCleanerRunOnceResult{QueuesDeleted: make([]string, 0, 10)}
+
+	for {
+		// Wrapped in a function so that defers run as expected.
+		queuesDeleted, err := func() ([]string, error) {
+			ctx, cancelFunc := context.WithTimeout(ctx, 30*time.Second)
+			defer cancelFunc()
+
+			queuesDeleted, err := s.exec.QueueDeleteExpired(ctx, &riverdriver.QueueDeleteExpiredParams{
+				Max:              s.batchSize,
+				UpdatedAtHorizon: time.Now().Add(-s.Config.QueueRetentionPeriod),
+			})
```

Comment:
Am I understanding correctly that when you use the global pause/resume of "*", you can never pause all queues for more than 24 hours?

Unlike a normal queue, River clients won't report on the global queue "*", so it won't have it's `updated_at` field bumped unless the user pauses/resumes again. So after 24 hours it gets reaped here even if the user never resumed an initial pause.

Reply:
That is certainly not the intention, no. The intention is that as long as `Client` is configured to work a queue and is actively started, the `updated_at` timestamp should be regularly bumped. The queue should never be deleted by the cleaner process unless no started Clients have been configured to use it within the past 24 hours.

The queries for pausing and resuming use a `CASE` to conditionally filter the `river_queue` rows by `name` only when `name != '*'`, so if `*` is the provided value, _all_ registered job queues are paused. However if you were to then start a new `Client` with a brand new queue name registered, it would not be paused because it was not in the table at the time you paused everything.

Can you do me a favor and specifically look at whether this is clear to you when you now look at the code? And if not, where might we improve that or better document it?

Reply:
Another way of stating this: calling `QueuePause` with `*` does not persist a queue record with the `*` value in it. Instead, it triggers the `CASE` in the pause query to avoid filtering queue rows by name and instead selects all unpaused queues for update. Those queues will continue to have their `updated_at` bumped periodically by any producer that is actively using them, so they will remain paused for as long as they are in use (plus the cleanup horizon).

Reply:
Ah, I see! Okay, my mistake — I misunderstood how that piece worked.